File: app/src/main/java/cz/johrusk/showsmscode/receiver/SmsReceiver.java
package cz.johrusk.showsmscode.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;

import cz.johrusk.showsmscode.service.MsgHandlerService;
import timber.log.Timber;

/**
 * Class which checks received SMS.
 *
 * @author Josef Hruska (pepa.hruska@gmail.com)
 */
public class SmsReceiver extends BroadcastReceiver {

    public Context c;

    @Override
    public void onReceive(Context c, Intent intent) {
        this.c = c;
        Timber.d("SMS RECEIVED");
        SmsMessage[] messages = getMessages(intent);
        for (SmsMessage message : messages) {
            try {
                Timber.d("receiving sms from " + message.getDisplayOriginatingAddress());
            } catch (Exception e) {
                e.printStackTrace();
            }

            if (message != null) {

                Intent msgHandler = new Intent(c, MsgHandlerService.class);
                String msgContent = message.getMessageBody();
                String msgSender = message.getDisplayOriginatingAddress();
                Bundle msg = new Bundle();

                Timber.d("Content of SMS " + msgContent + " / " + msgSender);
                msg.putStringArray("msg", new String[]{msgSender, msgContent});
                msgHandler.putExtra("msg", msg);
                c.startService(msgHandler);
            }
        }
    }

    public synchronized SmsMessage[] getMessages(Intent intent) {
        Bundle bundle = intent.getExtras();

        Object messages[] = (Object[]) bundle.get("pdus");
        if (messages != null) {
            try {
                SmsMessage[] smsMessages = new SmsMessage[messages.length];
                for (int i = 0; i < messages.length; i++) {
                    smsMessages[i] = SmsMessage.createFromPdu((byte[]) messages[i]);
                }
                return smsMessages;
            } catch (SecurityException e) {
                return new SmsMessage[0];
            }
        } else {
            return new SmsMessage[0];
        }
    }
}

File: app/src/main/java/cz/johrusk/showsmscode/receiver/SmsReceiver.kt
package cz.johrusk.showsmscode.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.telephony.SmsMessage
import cz.johrusk.showsmscode.service.MsgHandlerService
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.warn
import timber.log.Timber

/**
 * Created by Pepa on 03.07.2016.
 */
class SmsReceiver : BroadcastReceiver(), AnkoLogger {
    override fun onReceive(context: Context, intent: Intent) {
        warn { "SMS Received" }
        val messages = getMessages(intent)
        for (message in messages) {
            try {
                Timber.d("receiving sms from " + message!!.displayOriginatingAddress)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            if (message != null) {

                val msgHandler = Intent(context, MsgHandlerService::class.java)
                val msgContent = message.messageBody
                val msgSender = message.displayOriginatingAddress
                val msg = Bundle()

                Timber.d("Content of SMS $msgContent / $msgSender")
                msg.putStringArray("msg", arrayOf(msgSender, msgContent))
                msgHandler.putExtra("msg", msg)
                context.startService(msgHandler)
            }
        }
    }

    @Synchronized fun getMessages(intent: Intent): Array<SmsMessage?> {
        val bundle = intent.extras
        val format = bundle.getString("format") //From API 23+ it is necessary to get format to choose between 3GPP/3GPP2 formatting

        val messages: Array<Any>? = bundle.get("pdus") as Array<Any>
        if (messages != null) {
            try {
                val smsMessages = arrayOfNulls<SmsMessage>(messages.size)
                for (i in messages.indices) {
                    smsMessages[i] = SmsMessage.createFromPdu(messages[i] as ByteArray,format)
                }
                return smsMessages
            } catch (e: SecurityException) {
                return arrayOfNulls(0)
            }

        } else {
            return arrayOfNulls(0)
        }
    }
}